src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -144,14 +144,21 @@ private unsafe void UnsafeDeflate(ReadOnlySpan<byte> input, Span<byte> output, o
                 consumed = input.Length - (int)_stream.AvailIn;
                 written = output.Length - (int)_stream.AvailOut;
 
-                needsMoreBuffer = errorCode == ErrorCode.BufError || _stream.AvailIn > 0;
+                // It is important here to also check that we haven't
+                // exhausted the output buffer because after deflating we're
+                // always going to issue a flush and a flush with empty output
+                // is going to throw.
+                needsMoreBuffer = errorCode == ErrorCode.BufError
+                    || _stream.AvailIn > 0
+                    || written == output.Length;
```

Comment:
So we have tests that exercise all of these conditions?

Reply:
The first two yes. They are easy to be tested when we provide payload that is incompressible.

The `written == output.Length` I couldn't reproduce, but it is my only conclusion on how could an empty `Span<byte>` buffer reach `UnsafeFlush`.